src/platform/backends/shared/base_virtual_machine_factory.cpp

```diff
@@ -81,3 +82,28 @@ void mp::BaseVirtualMachineFactory::prepare_interface(NetworkInterface& net,
         }
     }
 }
+
+void mp::copy_instance_dir_with_essential_files(const fs::path& source_instance_dir_path,
+                                                const fs::path& dest_instance_dir_path)
+{
+    if (fs::exists(source_instance_dir_path) && fs::is_directory(source_instance_dir_path))
```

Comment:
Changing it to assert might make unit tests more complicated, because for now. It skips the if branch in the case of unit test. Anyway, I can give a try.

Reply:
As discussed, if a function has pre-conditions, unit tests need to fulfill them, one way or another. In this case, mockable file ops would probably be the right choice. So two changes: assert + mockable file ops.

Reply:
Hi @ricab 
About the unit test of this function, I have a [commit](https://github.com/canonical/multipass/pull/3752/commits/2b196564295eacab7e256091555e26890e2941be) for that based on the mocking file operations. However, I really do not think it is a good way to test that particular function because the setup (`is_exist`, `is_directory`, and also the constructor of `directory_iterator` throws if the actual path does not exist, by the way, I am aware that we have the `DirIterator` class) it requires. In this case, an integration test (using temporary directory) makes more sense to me. Regarding to the stability of that, we have problems with qt temporary directory, however, we also have `fs::temp_directory_path()` from std and we are using that in `test_file_ops.cpp` file. 

All in all, my personal opinion is that unit tests and integration tests should co-exist when it comes to testing file operations. The testing of this particular function further confirmed that.